From a0d06a696c770310841f28f6c0e71d71ee281613 Mon Sep 17 00:00:00 2001
From: Charles Keepax <ckeepax@opensource.cirrus.com>
Date: Thu, 6 Mar 2025 12:54:02 +0000
Subject: [PATCH 183/237] sof-soundwire: Split cs42l43 dmic initialisation

Split the initialisation of cs42l43 into a headset part and a dmic
part. This will prevent the dmic mute controls getting attached to
the mute LED when the codec dmics are not being used.

Link: https://lore.kernel.org/alsa-devel/20250306125402.1741562-1-ckeepax@opensource.cirrus.com/
Signed-off-by: Charles Keepax <ckeepax@opensource.cirrus.com>
Signed-off-by: Jaroslav Kysela <perex@perex.cz>
Origin: upstream, https://github.com/alsa-project/alsa-ucm-conf/commit/a0d06a696c770310841f28f6c0e71d71ee281613
Bug-Ubuntu: https://bugs.launchpad.net/bugs/2106394
Signed-off-by: Atlas Yu <atlas.yu@canonical.com>
---
 ucm2/codecs/cs42l43-dmic/init.conf    | 22 ++++++++++++++++++++++
 ucm2/codecs/cs42l43/init.conf         | 11 +----------
 ucm2/sof-soundwire/sof-soundwire.conf |  2 +-
 3 files changed, 24 insertions(+), 11 deletions(-)
 create mode 100644 ucm2/codecs/cs42l43-dmic/init.conf

--- /dev/null
+++ b/ucm2/codecs/cs42l43-dmic/init.conf
@@ -0,0 +1,22 @@
+# cs42l43 specific control settings
+
+LibraryConfig.remap.Config {
+	ctl.default.map {
+		"name='cs42l43 Microphone Capture Switch'" {
+			"name='cs42l43 Decimator 3 Switch'".vindex.0 0
+			"name='cs42l43 Decimator 4 Switch'".vindex.1 0
+		}
+		"name='cs42l43 Microphone Capture Volume'" {
+			"name='cs42l43 Decimator 3 Volume'".vindex.0 0
+			"name='cs42l43 Decimator 4 Volume'".vindex.1 0
+		}
+	}
+}
+
+FixedBootSequence [
+	cset "name='cs42l43 Decimator 3 Switch' 0"
+	cset "name='cs42l43 Decimator 4 Switch' 0"
+	exec "-/sbin/modprobe snd_ctl_led"
+	sysw "-/class/sound/ctl-led/mic/card${CardNumber}/attach:cs42l43 Decimator 3 Switch"
+	sysw "-/class/sound/ctl-led/mic/card${CardNumber}/attach:cs42l43 Decimator 4 Switch"
+]
--- a/ucm2/codecs/cs42l43/init.conf
+++ b/ucm2/codecs/cs42l43/init.conf
@@ -2,14 +2,6 @@
 
 LibraryConfig.remap.Config {
 	ctl.default.map {
-		"name='cs42l43 Microphone Capture Switch'" {
-			"name='cs42l43 Decimator 3 Switch'".vindex.0 0
-			"name='cs42l43 Decimator 4 Switch'".vindex.1 0
-		}
-		"name='cs42l43 Microphone Capture Volume'" {
-			"name='cs42l43 Decimator 3 Volume'".vindex.0 0
-			"name='cs42l43 Decimator 4 Volume'".vindex.1 0
-		}
 		"name='cs42l43 Headset Microphone Capture Switch'" {
 			"name='cs42l43 Decimator 1 Switch'".vindex.0 0
 		}
@@ -20,8 +12,7 @@
 }
 
 FixedBootSequence [
+        cset "name='cs42l43 Decimator 1 Switch' 0"
         exec "-/sbin/modprobe snd_ctl_led"
         sysw "-/class/sound/ctl-led/mic/card${CardNumber}/attach:cs42l43 Decimator 1 Switch"
-        sysw "-/class/sound/ctl-led/mic/card${CardNumber}/attach:cs42l43 Decimator 3 Switch"
-        sysw "-/class/sound/ctl-led/mic/card${CardNumber}/attach:cs42l43 Decimator 4 Switch"
 ]
--- a/ucm2/sof-soundwire/sof-soundwire.conf
+++ b/ucm2/sof-soundwire/sof-soundwire.conf
@@ -117,7 +117,7 @@
 If.mic_init {
 	Condition {
 		Type RegexMatch
-		Regex "(rt713-dmic|rt715(-sdca)?)"
+		Regex "(cs42l43-dmic|rt712-dmic|rt713-dmic|rt715(-sdca)?)"
 		String "${var:MicCodec1}"
 	}
 	True.Include.mic_init.File "/codecs/${var:MicCodec1}/init.conf"
